automation/terraform/testnets/gossipqa/main.tf

```diff
@@ -0,0 +1,273 @@
+
+  block_producer_key_pass           = "naughty blue worm"
+  block_producer_starting_host_port = 10501
+
+  snark_coordinators=[
+# gotta manually change the public keys, get them from whatever genesis ledger we're running this with
+# if we want to do this automatically, we would need to modify generate-keys-and-ledger.sh to create a separate batch of files with pulbic keys of all block producer nodes and read those in
+    {
+      snark_worker_replicas = 5
+      snark_worker_fee      = "0.025"
+      snark_worker_public_key = "B62qrqcDbsBpbWq8bCVA7xBa7Wp32CKVwZM6tkbRDxbEECbiy5epJo5"
+      snark_coordinators_host_port = 10401
+    },
+    {
+      snark_worker_replicas = 5
+      snark_worker_fee      = "0.025"
+      snark_worker_public_key = "B62qnzeVPJMgeWJpk13kRNoPzMLmTF1ZycLnEM6oZrAEu5VW2V9hftq"
+      snark_coordinators_host_port = 10402
+    }
+    # ,
+    # {
+    #   snark_worker_replicas = 5
+    #   snark_worker_fee      = "0.025"
+    #   snark_worker_public_key = "B62qnzeVPJMgeWJpk13kRNoPzMLmTF1ZycLnEM6oZrAEu5VW2V9hftq"
+    #   snark_coordinators_host_port = 10403
+    # },
+    # {
+    #   snark_worker_replicas = 5
+    #   snark_worker_fee      = "0.025"
+    #   snark_worker_public_key = "B62qnzeVPJMgeWJpk13kRNoPzMLmTF1ZycLnEM6oZrAEu5VW2V9hftq"
+    #   snark_coordinators_host_port = 10404
+    # },
+    # {
+    #   snark_worker_replicas = 5
+    #   snark_worker_fee      = "0.025"
+    #   snark_worker_public_key = "B62qnzeVPJMgeWJpk13kRNoPzMLmTF1ZycLnEM6oZrAEu5VW2V9hftq"
+    #   snark_coordinators_host_port = 10405
+    # },
+    # {
+    #   snark_worker_replicas = 5
+    #   snark_worker_fee      = "0.025"
+    #   snark_worker_public_key = "B62qnzeVPJMgeWJpk13kRNoPzMLmTF1ZycLnEM6oZrAEu5VW2V9hftq"
+    #   snark_coordinators_host_port = 10406
+    # }
+    # ,
+    # {
+    #   snark_worker_replicas = 5
+    #   snark_worker_fee      = "0.025"
+    #   snark_worker_public_key = "B62qnzeVPJMgeWJpk13kRNoPzMLmTF1ZycLnEM6oZrAEu5VW2V9hftq"
+    #   snark_coordinators_host_port = 10407
+    # },
+    # {
+    #   snark_worker_replicas = 5
+    #   snark_worker_fee      = "0.025"
+    #   snark_worker_public_key = "B62qnzeVPJMgeWJpk13kRNoPzMLmTF1ZycLnEM6oZrAEu5VW2V9hftq"
+    #   snark_coordinators_host_port = 10408
+    # },
+    # {
+    #   snark_worker_replicas = 5
+    #   snark_worker_fee      = "0.025"
+    #   snark_worker_public_key = "B62qnzeVPJMgeWJpk13kRNoPzMLmTF1ZycLnEM6oZrAEu5VW2V9hftq"
+    #   snark_coordinators_host_port = 10409
+    # },
+    # {
+    #   snark_worker_replicas = 5
+    #   snark_worker_fee      = "0.025"
+    #   snark_worker_public_key = "B62qnzeVPJMgeWJpk13kRNoPzMLmTF1ZycLnEM6oZrAEu5VW2V9hftq"
+    #   snark_coordinators_host_port = 10410
+    # }
+  ]
+
+  seed_count            = var.seed_count
+
+  plain_node_count = var.plain_node_count
+
+  whales= [    
+    {
+      duplicates = 2
+      class  = "whale"
+    },
+    {
+      duplicates = 1
+      class  = "whale"
```

Comment:
Can't you assume class = whale in the whales list? what is the class intended to be used for?

Reply:
in `o1-testnet/local.tf` the whale list and fish list actually get combined into a unified block_producer_configs, so the individual objects inside that combined list need to track what class they are a part of.  but yes i suppose this bit could be streamlined or have this aspect hidden from the top level of abstraction but i haven't bothered 😅

Reply:
The loops here: https://github.com/MinaProtocol/mina/pull/9284/files#diff-ddf135dd05aa30ee4bf1977dc50d574ea55ee861cc5f348284c9ea5131460dd5R43 should be responsible for setting the class in my opinion, given that the variables are named whale and fish explicitly. Alternatively you could keep the classes defined upstream and turn this into one loop that takes a list of block producers with their classes as opposed to 2 with nearly identical logic.

Reply:
oh yeah that's a good catch

Reply:
i moved it into the loop in o1-testnet/local.tf.  those loops are so damn messy but idk what to do about it